### DISCUSSION TITLE
Can’t use ‘Udev-symlinked’ USB port, while physical USB port ttyUSBx is working

### DISCUSSION THREAD

**Discussion by Code2702** [Koenkk/zigbee2mqtt#6612] [Devices]
Hi, I’m new to zigbee2mqtt, trying to install it in docker compose on RaspberryPi 4b. I already have eclipse-mosquitto and domoticz running in docker compose. 

Two devices are connected to the pi; a RFX433 and a ‘Slaesh's CC2652RB’ zigbee stick. Both are mapped via a (udev rule) symlink to the USB ports /dev/ttyUSB0 and /dev/ttyUSB1. 
”ls -l /dev/ “ gives:
crw-rw---- 1 root dialout 188,   0 Mar  5 16:59 ttyUSB0
crwxrwxrwx 1 root dialout 188,   1 Mar  5 16:59 ttyUSB1
lrwxrwxrwx 1 root root           7 Mar  5 16:59 ttyUSB_RFX433 -> ttyUSB0
lrwxrwxrwx 1 root root           7 Mar  5 16:59 ttyUSB_ZigbeeSlae -> ttyUSB1

I can start zigbee2mqtt successfully when using physical port /dev/ttyUSB1 to connect to the Slaesh zigbee stick (it was bound to ttyUSB1, sometimes it’s ttyUSB0). Logged in as root, the command “test -w /dev/ttyUSB_ZigbeeSlae && echo success || echo failure” gives: “**success**”. 

However, when I change the ttyUSB1 to: ‘port: /dev/ttyUSB_ZigbeeSlae’ in /data/configuration.yaml (making no other changes) I get an error. Log says:  Starting Zigbee2MQTT version 1.17.1 (commit #99274c7)/info  2021-03-05 20:21:34: Starting zigbee-herdsman (0.13.59) /error 2021-03-05 20:21:34: Error while starting zigbee-herdsman

I read topic #6505 , the zigbee2mqtt.io FAQ and zigbee2mqtt.io/information/zigbee2mqtt_fails_to_start, I already checked/tried the following:
- Instead of ‘/dev/ttyUSB_ZigbeeSlae’, I tried /dev/serial/by-id/ usb-Silicon_Labs_slae.sh_cc2652rb_stick_-_slaesh_s_iot_stuff_00_58_4H_40_47_CC_4B_6C-if00-port0   (I first removed the udev rule for /dev/ttyUSB_ZigbeeSlae)
- usermod -aG docker $USER
- usermod -a -G dialout $USER
- chmod 777 /dev/ttyUSB1 (slash zigbee stick was bound to ttyUSB1)
- chown $USER:dialout /dev/ttyUSB1
- I reverted to image: koenkk/zigbee2mqtt:1.17.1 instead of image: koenkk/zigbee2mqtt

I suspect it’s a permission issue or a bug, since “test -w /dev/ttyUSB_ZigbeeSlae && echo success || echo failure” is successful and docker is using ‘root’ as user as far as I know. This might be related to the problem described in #6607 ..

Anyone who can help solving this issue? I tried for hours but didn’t succeed. Your help would be greatly appreciated!

**Comment by Jim-Noble** [comment]
The problem is that docker doesn't like relative symbolic links. 
My solution was to create a udev rule that creates a symlink with a known name, then creates a hard link from the symlink. 
The latter just requires a couple of extra lines in the udev rule to "rm -f" the hard link and recreate it.

udev rules:
`SUBSYSTEM=="tty", ATTRS{idVendor}=="1a86", ATTRS{idProduct}=="7523", SYMLINK+="zzh"`
`SUBSYSTEM=="tty", ATTRS{idVendor}=="1a86", ATTRS{idProduct}=="7523", RUN+="/bin/rm -f /dev/zzhh"`
`SUBSYSTEM=="tty", ATTRS{idVendor}=="1a86", ATTRS{idProduct}=="7523", RUN+="/bin/ln -L /dev/zzh /dev/zzhh"`

You then specify "/dev/zzhh" (or whatever you want to call it) in your zigbee2mqtt config.

James

**Comment by Code2702** [comment]
Great solution, works perfectly! Thanks a lot James for helping me out!
